c2corg_api/views/document_revert.py

```diff
@@ -0,0 +1,236 @@
+        self.request = request
+
+    @restricted_json_view(
+        permission='moderator',
+        schema=RevertSchema(),
+        validators=[colander_body_validator, validate_version])
+    def post(self):
+        """ Create a new version of the document based upon an old one.
+
+        Request:
+            `POST` `/documents/revert`
+
+        Request body:
+            {
+                'document_id': @document_id@,
+                'lang': @lang@,
+                'version_id': @version_id@
+            }
+
+        """
+        document_id = self.request.validated['document_id']
+        lang = self.request.validated['lang']
+        version_id = self.request.validated['version_id']
+
+        clazz, locale_clazz, archive_clazz, \
+            archive_locale_clazz, document_type = self._get_models(document_id)
+
+        document = self._get_current_document(
+            document_id, lang, clazz, locale_clazz)
+        document_in = self._get_archive_document(
+            document_id, lang, version_id, archive_clazz, archive_locale_clazz)
+
+        old_versions = document.get_versions()
+        user_id = self.request.authenticated_userid
+
+        # update the document with the input document
+        document.update(document_in)
+
+        # now set back the versions to the versions of the current document
+        # because we basing our changes on the lastest version
+        document.version = old_versions['document']
+        document.locales[0].version = old_versions['locales'][lang]
+        if document.geometry:
+            document.geometry.version = old_versions['geometry']
+
+        # TODO? support 'before_update' callbacks as in _put()
+        # See 'after_update' handling below
+
+        try:
+            DBSession.flush()
+        except StaleDataError:
+            raise HTTPConflict('concurrent modification')
+
+        # when flushing the session, SQLAlchemy automatically updates the
+        # version numbers in case attributes have changed. by comparing with
+        # the old version numbers, we can check if only figures or only locales
+        # have changed.
+        (update_types, changed_langs) = document.get_update_type(old_versions)
+
+        if update_types:
+            # A new version needs to be created and persisted
+            message = 'Revert to version {}'.format(version_id)
+            DocumentRest.update_version(
+                document, user_id, message, update_types, changed_langs)
+
+            if document.type != AREA_TYPE and UpdateType.GEOM in update_types:
+                update_areas_for_document(document, reset=True)
+
+            if document.type != MAP_TYPE and UpdateType.GEOM in update_types:
+                update_maps_for_document(document, reset=True)
+
+            after_update = self._get_after_update(document_type)
+            if after_update:
+                after_update(document, update_types, user_id=user_id)
+
+            update_cache_version(document)
```

Comment:
You also have to notify ES:
https://github.com/c2corg/v6_api/blob/a300ced/c2corg_api/views/document.py#L299